File: kasku-backend/transaction-service/internal/delivery/http/router.go
package http

import (
	"github.com/TubagusAldiMY/kasku/observability-go/metrics"
	"github.com/TubagusAldiMY/kasku/transaction-service/internal/delivery/http/handler"
	"github.com/TubagusAldiMY/kasku/transaction-service/internal/delivery/http/middleware"
	"github.com/gin-gonic/gin"
	"github.com/rs/zerolog"
)

func NewRouter(h *handler.TransactionHandler, isDev bool, metricsReg *metrics.Registry, log zerolog.Logger) *gin.Engine {
	if !isDev {
		gin.SetMode(gin.ReleaseMode)
	}
	r := gin.New()
	r.Use(gin.Recovery())
	r.Use(middleware.CorrelationID())
	r.Use(metricsReg.HTTPMetrics())
	r.Use(securityHeaders())

	r.GET("/health", h.Health)
	r.GET("/metrics", gin.WrapH(metricsReg.Handler()))

	v1 := r.Group("/v1")
	{
		txs := v1.Group("/transactions")
		{
			txs.GET("", h.ListTransactions)
			txs.POST("", h.CreateTransaction)
			txs.GET("/export", h.ExportCSV)
			txs.GET("/:id", h.GetTransaction)
			txs.PUT("/:id", h.UpdateTransaction)
			txs.DELETE("/:id", h.DeleteTransaction)
		}
		cats := v1.Group("/categories")
		{
			cats.GET("", h.ListCategories)
			cats.POST("", h.CreateCategory)
			cats.PUT("/:id", h.UpdateCategory)
			cats.DELETE("/:id", h.DeleteCategory)
		}
		budgets := v1.Group("/budgets")
		{
			budgets.GET("", h.ListBudgets)
			budgets.POST("", h.CreateBudget)
			budgets.GET("/:id", h.GetBudget)
			budgets.PUT("/:id", h.UpdateBudget)
			budgets.DELETE("/:id", h.DeleteBudget)
		}
	}
	return r
}

func securityHeaders() gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Header("X-Content-Type-Options", "nosniff")
		c.Header("X-Frame-Options", "DENY")
		c.Header("X-XSS-Protection", "1; mode=block")
		c.Header("Referrer-Policy", "strict-origin-when-cross-origin")
		c.Next()
	}
}

File: kasku-backend-monolith/src/modules/transaction/delivery/router.rs
use std::sync::Arc;
use axum::{
    middleware,
    routing::{delete, get, post, put},
    Router,
};

use crate::app_state::AppState;
use crate::shared::middleware::auth::auth_middleware;
use crate::shared::middleware::tenant_inject::tenant_inject_middleware;
use crate::shared::middleware::tier_inject::tier_inject_middleware;
use super::handler;

pub fn transaction_router(state: Arc<AppState>) -> Router<Arc<AppState>> {
    Router::new()
        .route("/", get(handler::list_transactions).post(handler::create_transaction))
        .route("/summary", get(handler::get_summary))
        .route("/export", get(handler::export_csv))
        .route("/:id", get(handler::get_transaction).put(handler::update_transaction).delete(handler::delete_transaction))
        .route_layer(middleware::from_fn(tenant_inject_middleware))
        .route_layer(middleware::from_fn_with_state(
            state.clone(),
            tier_inject_middleware,
        ))
        .route_layer(middleware::from_fn_with_state(state.clone(), auth_middleware))
}

pub fn category_router(state: Arc<AppState>) -> Router<Arc<AppState>> {
    Router::new()
        .route("/", get(handler::list_categories).post(handler::create_category))
        .route("/:id", put(handler::update_category).delete(handler::delete_category))
        .route_layer(middleware::from_fn(tenant_inject_middleware))
        .route_layer(middleware::from_fn_with_state(state.clone(), auth_middleware))
}

pub fn budget_router(state: Arc<AppState>) -> Router<Arc<AppState>> {
    Router::new()
        .route("/", get(handler::list_budgets).post(handler::create_budget))
        .route("/:id", get(handler::get_budget).put(handler::update_budget).delete(handler::delete_budget))
        .route_layer(middleware::from_fn(tenant_inject_middleware))
        .route_layer(middleware::from_fn_with_state(
            state.clone(),
            tier_inject_middleware,
        ))
        .route_layer(middleware::from_fn_with_state(state.clone(), auth_middleware))
}
